### ISSUE TITLE
Compiler Bug: Inconsistent accessibility is not enforced

### ISSUE THREAD

**Issue by hemisphera** [microsoft/AL#7614]
**1. Describe the bug**
Inconsistent accessibility errors are not enforced compile-time, something like [CS0051](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs0051?f1url=%3FappId%3Droslyn%26k%3Dk(CS0051)) is checked for example in in C#.

An example would be an internal table that is exposed to the public via event publisher. Everyone can subscribe to a (normal integration or business) event, and therefore all parameters that this event exposes, must also be exposable.

The problem is that this can lead to potentially exposing things that one really didn't want to expose and - worse - the consumer obviously can't consume. Furthermore: one is then not even able to rectify because doing so would be a breaking change. So basically a whole lot of noise and mess and clean-up required of something that IMHO shouldn't event be allowed to compile.

**2. To Reproduce**

Create a random table

````AL
table 50100 ARandomTable
{
  Access = Internal;
  // do stuff
}
````

Create random codeunit with an event publisher.

````AL
codeunit 50100 ARandomCodeunit
{

    [IntegrationEvent(false, false)]
    local procedure DereBeEventsMon(var stuff: Record "ARandomTable")
    begin
    end;

}
````

**3. Expected behavior**
I would expect a compilation error as this violates accessibility constraints.

**4. Actual behavior**
It compiles just fine.

**5. Versions:**

- AL Language: v13.0.903630
- Visual Studio Code: N/A
- Business Central: N/A
- List of Visual Studio Code extensions that you have installed: N/A

**Comment by hemisphera** [comment]
Yes, the `internalsVisibleTo` is a valid concern. However, as you said, it's mostly used for unit testing apps - at least I think it should. As such you could easily limit the compiler error to arise only if no `internalsVisibleTo` has been specified. Of course you cannot know what app has what access, but when there is no app at all in that property, then my code example makes no sense.

Also: we have `InternalEvent` events. If I really intended my event to be consumed by an 'internals-enabled' app, I'd rather should be using that instead of a 'public' event, where internals are getting exposure. That would be way cleaner.

I get your point. But one of the biggest issues we face is removing something that was exposed by error because of the various breaking change policies. This would help in mitigation.

**Comment by ChrisBlankDe** [comment]
Another example for this would be an public table containing an public field of an internal enum type.
I would be very happy if we could check this at compile time and issue at least a warning

**Comment by JesperSchulz** [comment]
The fix for this issue has been checked in to the master branch. It will be available in the bcinsider.azurecr.io/bcsandbox-master Docker image starting from platform build number 24.0.15860.0 and VS Code Extension Version 13.0.935369. 
 
 If you don’t have access to these images you need to become part of the Ready2Go program: aka.ms/readytogo 
 
 For more details on code branches and docker images please read: 
 https://blogs.msdn.microsoft.com/nav/2018/05/03/al-developer-previews-multiple-releases-and-github/ 
 https://freddysblog.com/2020/06/25/working-with-artifacts/
